File: scripts/manager_core/note_aliases.py
"""Keep remote notes attached when an old catalog link becomes an editable ID.

Only metadata from the manager's existing SSH cache is used. No SSH connection,
record copy, or note-body rewrite is necessary.
"""
import hashlib
import json
from pathlib import Path
import threading
import uuid

from . import catalog_frames
from .store import atomic_json

_NAMESPACE = uuid.UUID('14a0f21b-b529-45fc-bd9e-b07637424fa3')
_LOCK = threading.RLock()


def document_path(root, task):
    encoded = json.dumps(task, ensure_ascii=False, separators=(',', ':')).encode('utf-8')
    return Path(root) / 'work/control-center/notes' / (hashlib.sha256(encoded).hexdigest() + '.json')


def refresh(root, task):
    host = task.get('host_id', '')
    thread = str(uuid.UUID(task['thread_id']))
    prefix = next((p for p in ('ssh:', 'remote-ssh-discovered:') if host.startswith(p)), None)
    if prefix is None or len(host) > 256 or any(ord(c) < 32 for c in host):
        raise ValueError('Invalid remote note host')
    current = dict(host_id=host, thread_id=thread)
    with _LOCK:
        if document_path(root, current).exists():
            return
        directory = Path(root) / 'work/control-center'
        cache = directory / 'catalog/ssh' / (hashlib.sha256(host[len(prefix):].encode()).hexdigest() + '.jsonl')
        if not cache.exists():
            return
        with cache.open('rb') as stream:
            data = catalog_frames.read(stream)
        candidates = set()
        for entry in data.get('conversations', []):
            if entry.get('thread_id') != thread:
                continue
            source = entry.get('source_store_id')
            if not isinstance(source, str) or not source or len(source) > 256:
                continue
            projection = str(uuid.uuid5(_NAMESPACE, f'local\0{source}\0{thread}'))
            previous = dict(host_id=host, thread_id=projection)
            if document_path(root, previous).is_file():
                candidates.add(projection)
        if len(candidates) > 1:
            raise ValueError('Several old task copies have notes; existing notes were preserved')
        if not candidates:
            return
        path = directory / 'note-aliases.json'
        if path.exists() and (path.is_symlink() or path.stat().st_size > 4 * 1024 * 1024):
            raise ValueError('Invalid note alias file')
        aliases = json.loads(path.read_text(encoding='utf-8')) if path.exists() else {}
        if not isinstance(aliases, dict) or len(aliases) >= 16384:
            raise ValueError('Invalid note alias map')
        key = host + '\0' + thread
        target = candidates.pop()
        if key in aliases and aliases[key] != target:
            raise ValueError('Remote note identity changed; existing notes were preserved')
        aliases[key] = target
        atomic_json(path, aliases)

File: manager/service/src/note_aliases.rs
//! Saved catalog links and editable remote IDs refer to the same note document.
use crate::notes::TaskKey;
use serde_json::Value;
use std::path::Path;

pub fn is_remote(host: &str) -> bool {
    host.starts_with("ssh:") || host.starts_with("remote-ssh-discovered:")
}

pub fn resolve(root: &Path, task: TaskKey) -> Result<TaskKey, String> {
    if !is_remote(&task.host_id) || task.path(root).exists() {
        return Ok(task);
    }
    let path = root.join("work/control-center/note-aliases.json");
    if !path.exists() {
        return Ok(task);
    }
    let metadata = std::fs::symlink_metadata(&path).map_err(|e| e.to_string())?;
    if !metadata.is_file() || metadata.len() > 4 * 1024 * 1024 {
        return Err("메모 연결 파일을 확인하지 못했습니다.".into());
    }
    let aliases: Value = serde_json::from_slice(&std::fs::read(path).map_err(|e| e.to_string())?)
        .map_err(|_| "메모 연결 정보가 손상되었습니다. 원본은 보존되었습니다.")?;
    let key = format!("{}\0{}", task.host_id, task.thread_id);
    let Some(target) = aliases.get(key).and_then(Value::as_str) else {
        return Ok(task);
    };
    if uuid::Uuid::parse_str(target).is_err() {
        return Err("메모 원본 작업 정보가 올바르지 않습니다.".into());
    }
    let resolved = TaskKey {
        host_id: task.host_id,
        thread_id: target.into(),
    };
    if !resolved.path(root).is_file() {
        return Err("연결된 원본 메모를 찾지 못했습니다. 새 메모로 덮어쓰지 않았습니다.".into());
    }
    Ok(resolved)
}
